vivisect/analysis/elf/elfplt.py

```diff
@@ -22,16 +66,99 @@ def analyze(vw):
+    If the two clash, Dynamics wins.
+
+    Linux binaries tend to have both Sections and Dynamics entries
+    OpenBSD is only Sections.
+    QNX has only Dynamics.
+    '''
+    global gots
+    filename = vw.getFileByVa(fileva)
+
+    gottup = gots.get(filename)
+    if gottup is not None:
+        return gottup
+
+    for va, size, name, fname in vw.getSegments():
+        if name == ".got" and fname == filename:
+            gotva = va
+            gotsize = size
+            break
+
+    # pull GOT info from Dynamics
+    fdyns = vw.getFileMeta(filename, 'ELF_DYNAMICS')
+    if fdyns is not None:
+        FGOT = fdyns.get('DT_PLTGOT')
+        if vw.getFileMeta(filename, 'addbase'):
+            imgbase = vw.getFileMeta(filename, 'imagebase')
+            logger.debug('Adding Imagebase: 0x%x', imgbase)
+            FGOT += imgbase
+        if gotva is not None:
+            if gotva != FGOT:
+                logger.warn("Dynamics and Sections have different GOT entries: S:0x%x D:0x%x. using Dynamics", gotva, FGOT)
+                # since Dynamics don't store the GOT size, just use to the end of the memory map
+                mmva, mmsz, mmperm, mmname = vw.getMemoryMap(FGOT)
+                moffset = gotva - mmva
+                gotsize = mmsz - moffset
+            gotva = FGOT
+
+    gots[filename] = (gotva, gotsize)
+    return gotva, gotsize
+
+def getPLTs(vw):
+    plts = []
+    pltva = None
+    # Thought:  This is DT_PLTGOT, although each ELF will/may have their own DT_PLTGOT.
+    for va, size, name, fname in vw.getSegments():
+        if name.startswith(".plt"):
+            pltva = va
+            pltsize = size
+            plts.append((pltva, pltsize))
+            break
+
+    # pull GOT info from Dynamics
+    for fname in vw.getFiles():
+        fdyns = vw.getFileMeta(fname, 'ELF_DYNAMICS')
+        if fdyns is not None:
+            FGOT = fdyns.get('DT_PLTREL')
+            newish = True
+            for pltva, pltsize in plts:
+                if FGOT == pltva:
+                    newish = False
+            if newish:
+                plts.append((FGOT, None))
+
+    return gots
 
 def analyzePLT(vw, ssva, ssize):
-        # make code for every opcode in PLT
+    try:
+        ''' 
+        analyze an entire section designated as "PLT" or "PLTGOT"
+        '''
+        emu = None
         sva = ssva
         nextseg = sva + ssize
-
         trampbr = None
         has_tramp = False
+        gotva, gotsize = getGOT(vw, ssva)
 
+        ###### make code for every opcode in PLT
+        # make and parse opcodes.  keep track of unconditional branches
         branchvas = []
+
+        # drag an emulator along to calculate branches, if available
+        try:
+            emu = vw.getEmulator()
+            emu.setRegister(e_i386.REG_EBX, gotva)  # every emulator will have a 4th register, and if it's not used, no harm done.
+        except Exception as e:
+            logger.debug("no emulator available: %r", e)
```

Comment:
we should probably break or return here if the emulator isn't available.

Reply:
yeah, it's inevitable.  i'm specifically throwing down here, saying "you must have an emulator to play ELFPLT!"